sdcflows/transform.py

```diff
@@ -732,18 +737,16 @@ def grid_bspline_weights(target_nii, ctrl_nii, dtype="float32"):
         coords[axis] = np.arange(sample_shape[axis], dtype=dtype)
 
         # Calculate the index component of samples w.r.t. B-Spline knots along current axis
+        # Size of locations is L
         locs = nb.affines.apply_affine(target_to_grid, coords.T)[:, axis]
-        knots = np.arange(knots_shape[axis], dtype=dtype)
-
-        distance = np.abs(locs[np.newaxis, ...] - knots[..., np.newaxis])
-        within_support = distance < 2.0
-        d_vals, d_idxs = np.unique(distance[within_support], return_inverse=True)
-        bs_w = cubic(d_vals)
 
-        colloc_ax = lil_array((knots_shape[axis], sample_shape[axis]), dtype=dtype)
-        colloc_ax[within_support] = bs_w[d_idxs]
+        # Size of knots is K + 6 so that all locations are fully covered by basis
+        knots = np.arange(-3, knots_shape[axis] + 3, dtype=dtype)
 
-        wd.append(colloc_ax)
+        # Our original design matrices had size (K, L)
+        # However, BSpline.design_matrix() generates a size of (L, K + 2),
+        # hence the trasposition (and zero-padding of 1 at every face when using these)
+        wd.append(BSpline.design_matrix(locs, knots, 3).T)
```

Comment:
Does it make sense to drop the extra knots, rather than pad the coefficients?

```suggestion
        wd.append(BSpline.design_matrix(locs, knots, 3)[:, 1:-1].T)
```

Reply:
Yeah, I thought about this, and it may be positive to add a frame of zeros in the coefficients map. That way we may avoid weird border effects when coefficients at the edge are very large.

Reply:
BTW - we probably can get rid of the padding problem by setting `extrapolate=True` when creating the design_matrix. I'm not 100% positive that is a great idea though.